### PR TITLE
chore: migrate config files to `pyproject.toml`

### PR DESCRIPTION
**Pull request by SauravMaheshkar**
This PR aims to drop the configuration files in favour of `pyproject.toml` in order to have a more minimal project structure.

Request for Review: @matthewdouglas

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -2,6 +2,30 @@
...
+column_limit = 88
+coalesce_brackets = true
+space_between_ending_comma_and_closing_bracket = true
+spaces_before_comment = 2
+split_before_bitwise_operator = true
+split_before_first_argument = true
+split_before_logical_operator = true
+split_before_named_assigns = true
+split_complex_comprehension = true
+
```
Comment (akx) [inline pyproject.toml:18]: This project uses `ruff-format` for formatting (see #1081), so this is unnecessary.

```suggestion
```

File: pyproject.toml [hunk 2]
```diff
@@ -2,6 +2,30 @@
...
+[tool.pytest.ini_options]
+addopts = "-rP"
```
Comment (akx) [inline pyproject.toml:20]: Maybe the commented-out `addopts` from the old INI file should be retained for documentation.

File: pyproject.toml [hunk 3]
```diff
@@ -2,6 +2,30 @@
...
+log_cli = true
+log_cli_level = "INFO"
+log_file = "logs/pytest.log"
+markers = [
+    "slow: mark test as slow",
+    "benchmark: mark test as a benchmark",
```
Comment (akx) [inline pyproject.toml:26]: These could be in alphabetical order, as they were in the INI file.

**Comment by matthewdouglas** [comment]
There's an older PR (#1078) that I had started a while ago for the same thing. I'm OK superseding that one with this PR but with some more changes to match the intent of #1078.

The biggest of those changes would be to move more from setup.py to pyproject.toml as was done in #1078.

**Comment by SauravMaheshkar** [comment]
> There's an older PR (#1078) that I had started a while ago for the same thing. I'm OK superseding that one with this PR but with some more changes to match the intent of #1078.
> 
> The biggest of those changes would be to move more from setup.py to pyproject.toml as was done in #1078.

@matthewdouglas I've ported over some of your work from #1078 to this PR. Before I port over changes to the Github Actions changes as well, let's resolve the older comments.

**Comment by SauravMaheshkar** [comment]
Gentle ping @matthewdouglas

File: pyproject.toml [hunk 4]
```diff
@@ -1,7 +1,93 @@
...
+    "numpy>=1.17"
+]
+
+[project.optional-dependencies]
+benchmark = ["pandas", "matplotlib"]
+docs = ["hf-doc-builder==0.5.0"]
+dev = [
+    "bitsandbytes[test]",
+    "build>=1.0.0,<2",
+    "ruff==0.3.2",
```
Comment (matthewdouglas) [inline pyproject.toml:51]:
```suggestion
    "ruff==0.6.9",
```
We're currently using a newer version; let's update.

**Review by matthewdouglas — APPROVED** [review]
Rebased on main and fixed a couple conflicts. Looks good otherwise. Thanks!

**Comment by akx** [comment]
docs/source/installation.mdx has outdated instructions now btw (see #1485).

**Pull request merged**